lightning-block-sync/src/lib.rs

```diff
@@ -124,3 +142,547 @@ pub struct BlockHeaderData {
+	/// ended up which may not be `new_header`. Note that iff the returned `Err` contains `Some`
+	/// header then the transition from `old_header` to `new_header` is valid.
+	async fn sync_listener<L: ChainListener, P: Poll>(
+		&mut self,
+		new_header: ValidatedBlockHeader,
+		old_header: &ValidatedBlockHeader,
+		chain_poller: &mut P,
+		chain_listener: &mut L,
+	) -> Result<(), (BlockSourceError, Option<ValidatedBlockHeader>)> {
+		let difference = self.find_difference(new_header, old_header, chain_poller).await
+			.map_err(|e| (e, None))?;
+		self.disconnect_blocks(difference.disconnected_blocks, chain_listener);
+		self.connect_blocks(
+			difference.common_ancestor,
+			difference.connected_blocks,
+			chain_poller,
+			chain_listener,
+		).await
+	}
+
+	/// Returns the changes needed to produce the chain with `current_header` as its tip from the
+	/// chain with `prev_header` as its tip.
+	///
+	/// Walks backwards from `current_header` and `prev_header`, finding the common ancestor.
+	async fn find_difference<P: Poll>(
+		&self,
+		current_header: ValidatedBlockHeader,
+		prev_header: &ValidatedBlockHeader,
+		chain_poller: &mut P,
+	) -> BlockSourceResult<ChainDifference> {
+		let mut disconnected_blocks = Vec::new();
+		let mut connected_blocks = Vec::new();
+		let mut current = current_header;
+		let mut previous = *prev_header;
+		loop {
+			// Found the common ancestor.
+			if current.block_hash == previous.block_hash {
+				break;
+			}
+
+			// Walk back the chain, finding blocks needed to connect and disconnect. Only walk back
+			// the header with the greater height, or both if equal heights.
+			let current_height = current.height;
+			let previous_height = previous.height;
+			if current_height <= previous_height {
+				disconnected_blocks.push(previous);
+				previous = self.look_up_previous_header(chain_poller, &previous).await?;
+			}
+			if current_height >= previous_height {
+				connected_blocks.push(current);
+				current = self.look_up_previous_header(chain_poller, &current).await?;
+			}
+		}
+
+		let common_ancestor = current;
+		Ok(ChainDifference { common_ancestor, disconnected_blocks, connected_blocks })
+	}
+
+	/// Returns the previous header for the given header, either by looking it up in the cache or
+	/// fetching it if not found.
+	async fn look_up_previous_header<P: Poll>(
+		&self,
+		chain_poller: &mut P,
+		header: &ValidatedBlockHeader,
+	) -> BlockSourceResult<ValidatedBlockHeader> {
+		match self.header_cache.look_up(&header.header.prev_blockhash) {
+			Some(prev_header) => Ok(*prev_header),
+			None => chain_poller.look_up_previous_header(header).await,
+		}
+	}
+
+	/// Notifies the chain listeners of disconnected blocks.
+	fn disconnect_blocks<L: ChainListener>(
+		&mut self,
+		mut disconnected_blocks: Vec<ValidatedBlockHeader>,
+		chain_listener: &mut L,
+	) {
+		for header in disconnected_blocks.drain(..) {
+			println!("Disconnecting block {}", header.block_hash);
```

Comment:
Good question. I'm wondering if I should drop the prints entirely and leave the onus on the caller to log inside their listeners. We're likely already doing that when connecting and disconnecting blocks, so the additional logging is probably redundant. It would save piping a logger through as well. Thoughts?

Reply:
Yea, that seems fine to me.